File: Jan2021/02Jan/ConvertSortedArrayToBinarySearchTree.java
/*
Problem Name: 108. Convert Sorted Array to Binary Search Tree
Problem Link: https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return create(nums, 0, nums.length-1);
    }
    static TreeNode create(int nums[], int l, int r){
        TreeNode root = null;
        if(l <= r){
            int mid = (l + r)/2;
            root = new TreeNode(nums[mid]);
            root.left = create(nums, l, mid-1);
            root.right = create(nums, mid+1, r);
        }
        return root;
    }
}

File: Jan2021/02Jan/ConvertSortedArrayToBinarySearchTree.cpp
/*
Problem name: Convert Sorted Array to Binary Search Tree
Problem link: https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return helper(nums, 0, nums.size());
    }
    TreeNode* helper(vector<int>& nums, int min, int max){
        if(min >= max)
            return NULL;
        int mid = (min+max)/2;
        TreeNode* node = new TreeNode(nums[mid]);
        node->left = helper(nums, min, mid);
        node->right = helper(nums, mid+1, max);
        return node;
    }
};
